### PR TITLE
Handle additional memory requests made by `RCOMBBASE`

### PR DESCRIPTION
**Pull request by Al-Kindi-0**
## Describe your changes
Updates the bus to handle the two memory requests made by `RCOMBBASE` and updates docs.

## Checklist before requesting a review
- Repo forked and branch created from `next` according to naming convention.
- Commit messages and codestyle follow [conventions](./CONTRIBUTING.md).
- Relevant issues are linked in the PR description.
- Tests added for new functionality.
- Documentation/comments updated according to changes.

### CODE REVIEW

File: air/src/trace/main_trace.rs [hunk 1]
```diff
@@ -92,19 +92,9 @@ impl MainTrace {
...
-    pub fn helper_1(&self, i: usize) -> Felt {
-        self.columns.get_column(DECODER_TRACE_OFFSET + USER_OP_HELPERS_OFFSET + 1)[i]
-    }
-
-    /// Third decoder helper register at row i.
-    pub fn helper_2(&self, i: usize) -> Felt {
-        self.columns.get_column(DECODER_TRACE_OFFSET + USER_OP_HELPERS_OFFSET + 2)[i]
+    /// j-th decoder helper register at `row`.
+    pub fn helper(&self, j: usize, i: usize) -> Felt {
+        self.columns.get_column(DECODER_TRACE_OFFSET + USER_OP_HELPERS_OFFSET + j)[i]
```
Comment (bobbinth) [inline air/src/trace/main_trace.rs:97]: nit: the comment uses `row` while the parameter is named `i` - I think we should use `row` for the parameter name as well.

Also, if we rename `i` to `row`, I would rename `j` to `i`.

File: docs/src/design/stack/crypto_ops.md [hunk 2]
```diff
@@ -178,4 +178,28 @@ After calling the `mem_stream ` with `x_ptr`, the operation does the following:
...
+
+The `RCOMBBASE` makes two memory access request. To simplify the description of these, we first define the following variables :
+
+$$
+v_1 = \sum_{i=0}^3\alpha_{i+5} \cdot h_{i}'
+$$
+
+$$
+v_2 = \sum_{i=0}^1\alpha_{i+5} \cdot h_{i + 4}'
+$$
```
Comment (bobbinth) [inline docs/src/design/stack/crypto_ops.md:191]: Are $h_i'$ and $h_{i+4}'$ correct here? I thought helper values come from the current row?

Also, for $v_2$ I think we are missing $\alpha_7$ and $\alpha_8$ terms.

File: processor/src/chiplets/aux_trace/mod.rs [hunk 3]
```diff
@@ -466,13 +468,56 @@ fn build_mstream_request<E: FieldElement<BaseField = Felt>>(
...
+
+    let factor2 = alphas[0]
+        + alphas[1].mul_base(Felt::from(op_label))
+        + alphas[2].mul_base(ctx)
+        + alphas[3].mul_base(a_ptr)
+        + alphas[4].mul_base(clk)
+        + alphas[5].mul_base(a0)
+        + alphas[6].mul_base(a1)
+        + alphas[7].mul_base(ZERO)
+        + alphas[8].mul_base(ZERO);
```
Comment (bobbinth) [inline processor/src/chiplets/aux_trace/mod.rs:510]: These two computations are very similar - I wonder if we should have a separate function for computing memory request values. For example, something like:
```Rust
fn compute_memory_read_request(alphas: &[E], ctx: Felt, clk: Felt, addr: Felt, value: Word) -> E
```
(or maybe `ctx` and `clk` could be replaced with `self` if we put this function on `MainTrace`).

This will make the intent of the code a bit more clear. Also, such a function may be useful in other contexts too.

**Review by bobbinth — CHANGES_REQUESTED** [review]
Thank you! Looks good! I rebased the PR from the latest code and left a few comments inline.

File: air/src/trace/main_trace.rs (hunk 1 above)
Comment (Al-Kindi-0) [reply air/src/trace/main_trace.rs:97]: Makes sense, I have updated `(i, j)` to `(row, i)`

File: docs/src/design/stack/crypto_ops.md (hunk 2 above)
Comment (Al-Kindi-0) [reply docs/src/design/stack/crypto_ops.md:191]: Yes, that's a typo. Fixed now!
For the remaining alpha terms, I have omitted those as they are multiplied by `ZERO`, should I include them and add a note stating the aforementioned fact?

File: processor/src/chiplets/aux_trace/mod.rs (hunk 3 above)
Comment (Al-Kindi-0) [reply processor/src/chiplets/aux_trace/mod.rs:510]: Agreed! I have now added the proposed method to `MainTrace`

File: air/src/trace/main_trace.rs [hunk 4]
```diff
@@ -367,6 +357,30 @@ impl MainTrace {
         self.columns.get_column(MEMORY_V_COL_RANGE.start + 3)[i]
     }
 
+    /// Computes a memory read request at `row` given randomness `alphas`, memory address `addr`
+    /// and value `value`.
+    pub fn compute_memory_read_request<E: FieldElement<BaseField = Felt>>(
```
Comment (plafer) [inline air/src/trace/main_trace.rs:362]: With this new function, we have 2 functions in the system that build a memory read request: this one, and `processor::chiplets::aux_trace::build_mem_request()`. We should have only one.

A few thoughts:

- I think `MainTrace` should only have dumb accessors. It shouldn't be computing complex functions of the trace columns.
- A memory read (for words) is always a random linear combination of `[op_label, ctx, addr, clk, VALUE]`. The issue we have is that `MLOAD*` and `MSTORE*` don't take these values from the same locations in the trace as `RCOMBBASE`. The solution IMO is to have `build_mem_request` take the values it needs as input, and caller can supply them by reading from the correct trace location.
- I would split `build_mem_request()` into 2: one that builds a memory access for a `Felt` (e.g. `build_mem_element_request()`), and another for a `Word` (e.g. `build_mem_word_request()`). This would help us think of them as separate requests, because they are (i.e. the virtual columns that they collapse into 1 are different).

In other words, we currently think of a memory request (as defined by `build_mem_request()`) as coming from `MLOAD*` or `MSTORE*`. This change would help us think of a memory request as a random linear combination of specific columns (depending on whether we're reading a `Felt` or `Word`), which can come from any operation (current and/or future ones).

File: processor/src/chiplets/aux_trace/mod.rs [hunk 5]
```diff
@@ -466,13 +468,34 @@ fn build_mstream_request<E: FieldElement<BaseField = Felt>>(
...
+    let tz1 = main_trace.helper(1, i);
+    let tzg0 = main_trace.helper(2, i);
+    let tzg1 = main_trace.helper(3, i);
+    let a0 = main_trace.helper(4, i);
+    let a1 = main_trace.helper(5, i);
+    let z_ptr = main_trace.stack_element(13, i);
+    let a_ptr = main_trace.stack_element(14, i);
+
+    let factor1 = main_trace.compute_memory_read_request(alphas, i, z_ptr, [tz0, tz1, tzg0, tzg1]);
+    let factor2 = main_trace.compute_memory_read_request(alphas, i, a_ptr, [a0, a1, ZERO, ZERO]);
```
Comment (plafer) [inline processor/src/chiplets/aux_trace/mod.rs:487]: Given previous comment, this would call the new `build_mem_word_request()`

File: air/src/trace/main_trace.rs [hunk 6]
```diff
@@ -367,6 +357,30 @@ impl MainTrace {
...
+        alphas: &[E],
+        row: usize,
+        addr: Felt,
+        value: Word,
+    ) -> E {
+        let op_label = MEMORY_READ_LABEL;
+        let ctx = self.ctx(row);
+        let clk = self.clk(row);
+
+        alphas[0]
```
Comment (plafer) [inline air/src/trace/main_trace.rs:373]: Just a note: with #1223 we would take `alphas[0]` out of each memory request (such that each request is truly a random linear combination as opposed to an affine one). But we should leave as is in this PR, and change it only in #1223.

**Review by plafer — CHANGES_REQUESTED** [review]
Left a change proposal. Would love to hear your thoughts!

File: air/src/trace/main_trace.rs (hunk 4 above)
Comment (bobbinth) [reply air/src/trace/main_trace.rs:362]: > With this new function, we have 2 functions in the system that build a memory read request: this one, and `processor::chiplets::aux_trace::build_mem_request()`.

Ah - good point! I forgot about this one. We actually also have `build_mstream_request()` which does something similar. I'd love to consolidate them to use the same logic for computing the random linear combination.

> I think `MainTrace` should only have dumb accessors. It shouldn't be computing complex functions of the trace columns.

I agree with this. I wanted to have a more "convenient" location for the function - but separation of concerns is probably more important here.

As mentioned above, I still think that it would be good to have the function that computes the random linear combination for the request - but this function could probably live in the `aux_trace` module. Than, this function could be used from `build_mem_request()`, `build_mstream_request()`, `build_rcomb_base_request()` and others in the future.

> I would split `build_mem_request()` into 2: one that builds a memory access for a `Felt` (e.g. `build_mem_element_request()`), and another for a `Word` (e.g. `build_mem_word_request()`). This would help us think of them as separate requests, because they are (i.e. the virtual columns that they collapse into 1 are different).

If we have the above-mentioned function as a helper, I think splitting up `build_mem_request()` would make sense.

File: air/src/trace/main_trace.rs [hunk 7]
```diff
@@ -92,19 +92,9 @@ impl MainTrace {
...
-    /// Second decoder helper register at row i.
-    pub fn helper_1(&self, i: usize) -> Felt {
-        self.columns.get_column(DECODER_TRACE_OFFSET + USER_OP_HELPERS_OFFSET + 1)[i]
-    }
-
-    /// Third decoder helper register at row i.
-    pub fn helper_2(&self, i: usize) -> Felt {
-        self.columns.get_column(DECODER_TRACE_OFFSET + USER_OP_HELPERS_OFFSET + 2)[i]
+    /// The i-th decoder helper register at `row`.
+    pub fn helper(&self, i: usize, row: usize) -> Felt {
```
[the remark this answers is not in the document]
Comment (Al-Kindi-0) [reply air/src/trace/main_trace.rs]: Yes, makes sense. Renamed

File: processor/src/chiplets/aux_trace/mod.rs (hunk 5 above)
Comment (Al-Kindi-0) [reply processor/src/chiplets/aux_trace/mod.rs]: Makes sense, changed!

File: air/src/trace/main_trace.rs (hunk 4 above)
Comment (Al-Kindi-0) [reply air/src/trace/main_trace.rs]: I agree with points 1 and 2. Regarding point 3, it is indeed helpful (especially for the future) to make the code reflect the separation between requests (and by transitivity virtual columns). The issue is that this can be in conflict with reducing the size of the code in some places. Nevertheless, I have split the `build_mem_request()` and we can discuss the other cases, if any, on a case-by-case basis

File: processor/src/chiplets/aux_trace/mod.rs [hunk 8]
```diff
@@ -922,3 +930,27 @@ fn addr_to_hash_cycle(addr: Felt) -> usize {
 fn addr_to_row_index(addr: Felt) -> usize {
     (addr.as_int() - 1) as usize
 }
+
+/// Computes a memory read or write request at `row` given randomness `alphas`, memory address
+/// `addr` and value `value`.
+pub fn compute_memory_request<E: FieldElement<BaseField = Felt>>(
```
Comment (plafer) [inline processor/src/chiplets/aux_trace/mod.rs:936]:
```suggestion
fn compute_memory_request<E: FieldElement<BaseField = Felt>>(
```

File: processor/src/chiplets/aux_trace/mod.rs [hunk 9]
```diff
@@ -382,126 +384,129 @@ fn build_bitwise_request<E: FieldElement<BaseField = Felt>>(
...
-            main_trace.helper_1(i),
-            main_trace.helper_2(i),
-            main_trace.stack_element(0, i + 1),
-        )
-    };
+    let word = [
+        main_trace.stack_element(0, row + 1),
+        main_trace.helper_register(2, row),
+        main_trace.helper_register(1, row),
+        main_trace.helper_register(0, row),
```
Comment (plafer) [inline processor/src/chiplets/aux_trace/mod.rs:398]: What is stored in these helper registers, and why is it included as part of the request computation? Can we add a comment that explains it?

This would help understand the `value` argument to `compute_memory_request()`, which is currently unclear to me (since it's not always "the word that was read or written").

And based on this new understanding, I might rename `build_mem_request_{element, word}()`.

Comment (plafer) [reply processor/src/chiplets/aux_trace/mod.rs:398]: I infer from `build_memory_chiplet_response()` that the helper registers must hold the remaining 3 elements that were read/written. 

I would 

- make that clear in a comment here
- In `compute_memory_request()`'s docstring, I would add a sentence to say that `value` is the memory value that was read/written
- rename `build_mem_request_element()` -> `build_loadstore_request()`
- rename `build_mem_request_word()` -> `build_loadstorew_request()`

I suggest renaming `build_mem_request_{element, word}()` yet again because I had something else in mind when I suggested it. I thought we would have these 2 basic functions that everything else would depend on (which is now `compute_memory_request()`). Now these 2 are associated exclusively with the `MLOAD/STORE{W}` instructions, and so I would rename them so as to capture that.

File: air/src/trace/main_trace.rs [hunk 10]
```diff
@@ -92,19 +92,9 @@ impl MainTrace {
...
-    pub fn helper_1(&self, i: usize) -> Felt {
-        self.columns.get_column(DECODER_TRACE_OFFSET + USER_OP_HELPERS_OFFSET + 1)[i]
-    }
-
-    /// Third decoder helper register at row i.
-    pub fn helper_2(&self, i: usize) -> Felt {
-        self.columns.get_column(DECODER_TRACE_OFFSET + USER_OP_HELPERS_OFFSET + 2)[i]
+    /// The i-th decoder helper register at `row`.
+    pub fn helper_register(&self, i: usize, row: usize) -> Felt {
+        self.columns.get_column(DECODER_TRACE_OFFSET + USER_OP_HELPERS_OFFSET + i)[row]
```
Comment (bobbinth) [inline air/src/trace/main_trace.rs:97]: A thought for the future: we should probably introduce a newtype for to represent row. Maybe we call it `Cycle` or `Step`. It could look something like this:
```
pub struct Step(u32);
```

And then we can implement both `From<Step> for u32` and `From<Step> for usize`.

Let's create an issue for this.

**Review by bobbinth — APPROVED** [review]
Looks good! Thank you! I left one comment for the future inline.

**Review by plafer — APPROVED** [review]
Looks good :)
